handle pod roles starting with /

roffe opened the pull request:
Should be 100% backwards compatible. builded and tested in my staging cluster and it accepts pod roles with leading / and does what i expect it to :)

Closes https://github.com/uswitch/kiam/issues/75

pingles commented:
@roffe thanks for adding this- could you add some tests to `test/unit/arn_resolver_test.go` to demonstrate the new behaviour pleaes?

roffe commented:
There we go, Now with 3 tests added

Test 1: Add a role that begins with a /
Test 2: Add a role with a path not beginning with slash
Test 3: Add a role with a path beggining with slash

Hope this satisfies what you asked for @pingles

roffe commented:
ok, need to put some more work in, after i moved it from the pod_cache to the arn_resolver i'm getting this error when running in my staging cluster

```
{"addr":"100.96.9.14:39726","level":"error","method":"GET","msg":"error processing request: assume role forbidden: requested 'staging.sgn.io_seacucumber-serve' but annotated with '/staging.sgn.io_seacucumber-serve', forbidden","path":"/latest/meta-data/iam/security-credentials/staging.sgn.io_seacucumber-serve","status":403,"time":"2018-06-04T16:44:36Z"}
```

pingles commented:
ah- there’s policy which checks whether regexes permit etc. There’s also
one that the annotation matches what was requested- it’s probably necessary
to change that to use the resolver, so both sides use the resolved value.

On Mon, 4 Jun 2018 at 17:49, Joakim Karlsson <notifications@github.com>
wrote:

> ok, need to put some more work in, after i moved it from the pod_cache to
> the arn_resolver i'm getting this error when running in my staging cluster
>
> {"addr":"100.96.9.14:39726","level":"error","method":"GET","msg":"error processing request: assume role forbidden: requested 'staging.sgn.io_seacucumber-serve' but annotated with '/staging.sgn.io_seacucumber-serve', forbidden","path":"/latest/meta-data/iam/security-credentials/staging.sgn.io_seacucumber-serve","status":403,"time":"2018-06-04T16:44:36Z"}
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/uswitch/kiam/pull/80#issuecomment-394423186>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAAEfpKcypM8iKG5beSADV9FHZIbuJkiks5t5WUOgaJpZM4UZKfo>
> .
>

roffe commented:
built `roffe/kiam:issue-75-4` for the last commit b03c3f8 and i'm running it in my staging cluster now

Added test cases in all combinations i could think of and made so the role is passed through the resolver for sanitazion in policy.go

I hope this is at least is close to what you had envisioned :)

pingles commented:
Thanks @roffe. I've taken a look through- given the policies are now using the resolvers, perhaps it'd be cleaner/safer for the policies to be constructed with a resolver and then the server can pass it's configured resolver. I feel like without that it's possible for the annotated policy to differ from what the resolver interprets?

Other than that it looks good (and I appreciate you adding more tests 😄 )

roffe commented:
yeah that makes sense, i was taking some shortcuts with `ARNresolver := sts.DefaultResolver("")` as i wasn't sure about how you felt about doing said incision. I'll see what i can do! :)

pingles commented:
Thanks @roffe. If not I'll see if I can pick up from where you've got to.

roffe commented:
i removed the part i added from NamespacePermittedRoleNamePolicy since it's regex matching. Using the configured resolver makes no sense, nor does it to actually strip anything here, this is on the users end to configure their regex properly allow the name they actually put in the role.

I kept the tests / added for it and updated the test regext to allow the role with or without a leading slash

roffe commented:
First impression is it runs fine in my staging cluster so far :)

pingles commented:
Changes look good to me. It highlights that `Resolve` never returns an error so I'll refactor that interface tomorrow to tidy things up before I merge.

Thanks @roffe
